---
title: "edgeR"
author: "Katarzyna Sikora"
output: html_document
---

```{r setup, include=FALSE}
wd<-file.path(snakemake@params[["basedir"]],snakemake@params[["outdir"]])
system(paste0('mkdir -p ',wd))
knitr::opts_chunk$set(echo = TRUE)
knitr::opts_knit$set(root.dir = wd)
```

## Libs

```{r libs}
library(tximport)
library(edgeR)
library(GenomicRanges)
```

## Import counts and summarize gene-wise

Original oarfish output will be used.

The  output files are read using tximport function, which imports the transcript-level counts and summarizes them per gene using a transcript to gene mapping file.

```{r tximport}
input_files<-unlist(snakemake@params[["input_files"]])
names(input_files)<-gsub(".quant","",basename(input_files))
tx2gene_file<-snakemake@input[["t2g"]]
tx2gene<-read.delim(tx2gene_file,header=FALSE)[,c(1,2)]
txi<-tximport(input_files, type = "oarfish", tx2gene = tx2gene)

```

## Prep the DGEList object


```{r sampleinfo}
data.couts<-txi$counts
write.table(data.counts,"data.counts.tsv",sep="\t",quote=FALSE)

sampleSheet<-snakemake@input[["sampleSheet"]]
sampleInfo<-read.table(sampleSheet,sep="\t",header=TRUE)
rownames(sampleInfo)<-sampleInfo$name
write.table(sampleInfo,"sampleInfo.tsv",sep="\t",quote=FALSE)
data.counts<-data.counts[,sampleInfo$name]


y <- DGEList(counts = data.counts, samples = sampleInfo, group = sampleInfo$condition, genes = data$annotation)

head(y$genes)

#gtf_file<-snakemake@input[["gtf_file"]]
#gtf<-rtracklayer::import(gtf_file)
head(y$genes)
```


## Filter and normalize

Lowly expressed genes are filtered out prior to the downstream analysis.   

Scaling factors can computed using the TMM method to convert the resulting library sizes to effective library sizes.   

```{r filt norm}
keep <- filterByExpr(y)
table(keep)
y <- y[keep, , keep.lib.sizes=FALSE]


y <- normLibSizes(y)
y$samples
```


## Calculate MDS

MDS plots can also be used to visualize differences between the expression profiles of different samples with gene-level counts.   

```{r MDS}
plotMDS(y,col = c(1:2)[y$samples$condition],labels = y$samples$name,xlim = c(-4,4))
```

## Design matrix

We create the design matrix to compare HEK293 cells against HAP1 cells.   

```{r design}
design <- model.matrix(~ condition, data = y$samples)
design
```


## Dispersion estimation

Estimate and visualize NB dispersions.   

```{r disp}
y <- estimateDisp(y, design, robust=TRUE)
saveRDS(y,"y.RDS")
y$common.dispersion
plotBCV(y)

```

The NB dispersion estimates will not be used further under the latest quasi-likelihood (QL) pipeline.   
For DGE analyses, we're going to use the quasi-likelihood (QL) pipeline for stricter error rate control by accounting for the uncertainty associated with the dispersion estimation.

```{r quasil}
fit <- glmQLFit(y, design, robust=TRUE)
plotQLDisp(fit)
```

## Differential expression

Differentially expressed genes are tested between cell lines using the QL F-test.

```{r dte}
qlf <- glmQLFTest(fit)
is.de <- decideTests(qlf, p.value=0.05)
summary(is.de)


tt <- as.data.frame(topTags(qlf, n = Inf,p.value=0.05))
head(tt)
length(unique(tt$gene_id))
```
## MA plot

```{r ma}
plotMD(qlf)
```

## Save and export results

```{r save}
saveRDS(qlf,"qlf.RDS")
write.table(tt,"topTags_pval0.05.tsv",sep="\t",quote=FALSE)

```


## Session Info

```{r session info}
sessionInfo()
```


